File: 2025/go/benchmark2/parse_sum.go
package main

import (
	"encoding/binary"
	"fmt"
	"os"
	"time"
)

func parseAndSum(filePath string) (int64, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return 0, err
	}
	defer file.Close()

	var buffer []byte
	stat, err := file.Stat()
	if err != nil {
		return 0, err
	}
	buffer = make([]byte, stat.Size())

	_, err = file.Read(buffer)
	if err != nil {
		return 0, err
	}

	var sum int64
	for i := 0; i < len(buffer); i += 4 {
		// Convert 4 bytes to a little-endian 32-bit integer
		num := int32(binary.LittleEndian.Uint32(buffer[i : i+4]))
		sum += int64(num)
	}

	return sum, nil
}

func main() {
	filePath := "data.bin" // File path for the large binary file
	start := time.Now()

	result, err := parseAndSum(filePath)
	if err != nil {
		fmt.Println("Error:", err)
		return
	}

	elapsed := time.Since(start)
	fmt.Printf("Sum of integers: %d\n", result)
	fmt.Printf("Execution time (Go): %.4f seconds\n", elapsed.Seconds())
}

File: 2025/go/benchmark2/parse_sum.rs
use std::fs::File;
use std::io::{self, Read};
use std::time::Instant;

fn parse_and_sum(file_path: &str) -> io::Result<i64> {
    let mut file = File::open(file_path)?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;

    // Each 4 bytes represents a 32-bit integer
    let mut sum: i64 = 0;
    for chunk in buffer.chunks_exact(4) {
        let num = i32::from_le_bytes([chunk[0], chunk[1], chunk[2], chunk[3]]);
        sum += num as i64;
    }

    Ok(sum)
}

fn main() -> io::Result<()> {
    let file_path = "data.bin"; // File path for the large binary file
    let start = Instant::now();

    let result = parse_and_sum(file_path)?;
    let elapsed = start.elapsed();

    println!("Sum of integers: {}", result);
    println!("Execution time (Rust): {:.4} seconds", elapsed.as_secs_f64());

    Ok(())
}
